File: backend/app/schemas/migration.py
from datetime import datetime
from typing import Any, Literal

from app.llm.schemas import MigrationPlan
from pydantic import BaseModel, ConfigDict, model_validator


class MigrationTarget(BaseModel):
    """迁移目标框架和语言。"""

    framework: Literal["vue3"] = "vue3"
    language: Literal["typescript"] = "typescript"
    component_style: Literal["script_setup"] = "script_setup"


class MigrationCreateRequest(BaseModel):
    """前端请求创建迁移任务时的请求体。"""

    # React 文件名
    filename: str

    # React 组件源代码
    source_code: str

    target: MigrationTarget


class MigrationResponse(BaseModel):
    """返回给前端的迁移任务详情。"""

    # 允许 Pydantic 直接读取 SQLAlchemy ORM 对象属性。
    model_config = ConfigDict(from_attributes=True)

    # 迁移任务唯一标识。
    # 用于查询状态以及恢复 LangGraph 工作流。
    id: int

    # 用户提交的 React 文件名。
    filename: str

    # 当前任务状态。
    # 例如 queued、analyzing、waiting_for_review、
    # validating、completed、failed。
    status: str

    # 当前正在执行的 LangGraph 节点。
    current_node: str | None = None

    # 任务创建时间。
    created_at: datetime

    # 任务最后更新时间。
    updated_at: datetime | None = None

    # AI 生成、等待用户审核的迁移计划。
    migration_plan: MigrationPlan | None = None

    # 用户确认或修改后的最终迁移计划。
    approved_plan: MigrationPlan | None = None

    # 用户审核时填写的反馈。
    review_feedback: str | None = None

    # 最终生成的 Vue 文件名。
    generated_filename: str | None = None

    # 最终生成的 Vue 代码。
    generated_code: str | None = None

    # AI 生成代码时返回的补充说明。
    generation_notes: list[str] | None = None

    # SFC、Lint、类型检查、Build 是否全部通过。
    validation_passed: bool | None = None

    # 完整代码验证结果。
    validation_result: dict[str, Any] | None = None

    # 最终迁移报告。
    migration_report: dict[str, Any] | None = None


class MigrationDetailResponse(MigrationResponse):
    """迁移任务详情。"""

    # 用户驳回迁移计划时填写的原因
    review_feedback: str | None = None


class MigrationReviewRequest(BaseModel):
    """人工审核迁移计划时提交的数据。"""

    action: Literal["approve", "request_changes", "cancel"]

    # 通过时提交最终确认的计划
    migration_plan: MigrationPlan | None = None

    # 要求修改时填写具体反馈
    feedback: str | None = None

    @model_validator(mode="after")
    def validate_review_content(self) -> "MigrationReviewRequest":
        if self.action == "approve" and self.migration_plan is None:
            raise ValueError("通过审核时 migration_plan 不能为空")

        if self.action == "request_changes" and not (self.feedback or "").strip():
            raise ValueError("要求修改计划时 feedback 不能为空")

        return self

File: frontend/src/types/migration.ts
/**
 * 后端数据库中的任务状态。
 *
 * queued 是后续建议使用的初始状态。
 */
export type MigrationStatus =
  | "queued"
  | "analyzing"
  | "analyzed"
  | "planning"
  | "planned"
  | "waiting_for_review"
  | "approved"
  | "revision_requested"
  | "revising_plan"
  | "cancelled"
  | "rejected"
  | "generating"
  | "generated"
  | "validating"
  | "validated"
  | "validation_failed"
  | "repairing"
  | "repaired"
  | "report_generated"
  | "completed"
  | "failed";

/**
 * 页面状态。
 *
 * idle、submitting 只存在于前端；
 * 其他状态来自后端。
 */
export type MigrationPhase = "idle" | "submitting" | MigrationStatus;

/**
 * 工作流步骤的显示状态。
 */
export type WorkflowStepState = "pending" | "active" | "done" | "error";

/**
 * 迁移目标配置。
 */
export interface MigrationTarget {
  framework: "vue3";
  language: "typescript";
  component_style: "script_setup";
}

/**
 * POST /api/migrations/ 请求参数。
 */
export interface MigrationCreatePayload {
  filename: string;
  source_code: string;
  target: MigrationTarget;
}

/**
 * 单条 React → Vue 迁移映射。
 */
export interface MigrationMapping {
  category:
    | "state"
    | "effect"
    | "props"
    | "event"
    | "template"
    | "dependency"
    | "other";
  source: string;
  target: string;
  reason: string;
  source_line: number | null;
  strategy: "rule" | "llm";
  confidence: number;
  needs_review: boolean;
}

/**
 * 等待人工审核的结构化迁移计划。
 */
export interface MigrationPlan {
  component_name: string;
  target_framework: "vue3";
  target_language: "typescript";
  component_style: "script_setup";
  mappings: MigrationMapping[];
  dependencies: string[];
  risks: string[];
  manual_checks: string[];
  generation_constraints: string[];
}

/**
 * 单条 Vue 工具链验证错误。
 */
export interface ValidationIssue {
  stage:
    | "sfc_parse"
    | "script_compile"
    | "template_compile"
    | "migration_rule"
    | "lint"
    | "type_check"
    | "build"
    | "runner";
  message: string;
  line: number | null;
  column: number | null;
}

/**
 * SFC、Lint、Type 和 Build 的完整验证结果。
 */
export interface MigrationValidationResult {
  success: boolean;
  checks: Record<string, boolean>;
  errors: ValidationIssue[];
  details: Record<string, string>;
}

export interface MigrationReportSummary {
  component_name: string;
  source_filename: string;
  generated_filename: string;
  result: string;
  repair_count: number;
}

export interface MigrationReportStatistics {
  mapping_count: number;
  rule_mapping_count: number;
  llm_mapping_count: number;
  needs_review_count: number;
}

/**
 * 后端生成的结构化迁移报告。
 */
export interface MigrationReport {
  version: string;
  generated_at: string;
  summary: MigrationReportSummary;
  statistics: MigrationReportStatistics;
  mappings: MigrationMapping[];
  validation: Omit<MigrationValidationResult, "success"> & { passed: boolean };
  risks: string[];
  manual_checks: string[];
  generation_notes: string[];
  markdown: string;
}

/**
 * POST /api/migrations/{task_id}/review 请求参数。
 */
export type MigrationReviewPayload =
  | {
      action: "approve";
      migration_plan: MigrationPlan;
    }
  | {
      action: "request_changes";
      feedback: string;
    }
  | {
      action: "cancel";
    };

/**
 * POST 创建接口和 GET 查询接口返回的任务结构。
 */
export interface MigrationTask {
  id: number;
  filename: string;
  status: MigrationStatus;
  current_node: string | null;
  created_at: string;
  updated_at?: string | null;
  migration_plan?: MigrationPlan | null;
  approved_plan?: MigrationPlan | null;
  review_feedback?: string | null;
  generated_filename?: string | null;
  generated_code?: string | null;
  generation_notes?: string[] | null;
  validation_passed?: boolean | null;
  validation_result?: MigrationValidationResult | null;
  migration_report?: MigrationReport | null;
}
